csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,224 @@ class RedundantUseAnalysis : BackwardVisitor {
 
 } // namespace
 
+// This function removes the redundant write to gmem when an output tensor has a
+// leaf domain merged from concretized broadcast root domain and parallelized by
+// thread/block id. issue https://github.com/csarofeen/pytorch/issues/2125
+void ThreadPredicateMap::avoidConcretizedBroadcastRedundantWrite(
+    const TensorView* out_tv) {
+  // step-1, check if there is a leaf domain parallelized by thread/block id and
+  // its definition is a merged
+  bool merged_parallelized_thread_block = false;
+  for (auto ld : out_tv->domain()->domain()) {
+    const ParallelType& pt = ld->getParallelType();
+    auto merge = dynamic_cast<Merge*>(ld->definition());
+    if (isParallelTypeThread(pt) && merge) {
+      merged_parallelized_thread_block = true;
+      break;
+    }
+  }
+  // shortcut if there is no such leaf domain
+  if (!merged_parallelized_thread_block) {
+    return;
+  }
+
+  const auto& root_domain = out_tv->getMaybeRFactorDomain();
+  // For each broadcast root domain, find a concretized domain from its
+  // exact mapped domain set.
+  auto getConcretizedBroadcastRootDomain = [&]() {
+    std::unordered_map<IterDomain*, IterDomain*>
+        concretized_broadcast_root_domains;
+    for (auto rd : root_domain) {
+      if (rd->isBroadcast()) {
+        const auto& mapped_rd_exact_set = GpuLower::current()
+                                              ->caMap()
+                                              ->getIdSets(IdMappingMode::EXACT)
+                                              .getDisjointSetOf(rd);
+        for (auto mapped_rd : mapped_rd_exact_set) {
+          std::unordered_set<IterDomain*> all_cids =
+              GpuLower::current()
+                  ->concretizedBroadcastDomains()
+                  ->allConcretizedDomains(mapped_rd);
```

Comment:
I think you should assert that all the c_ids returned by this are in the same group in the `IdMappingMode::EXACT` graph.

Reply:
added check.

Reply:
Not sure why this assertion should hold. A broadcast ID can be concretized to different non-broadcast IDs and they don't need to be exactly mapped. They should be only permissively mapped.

Reply:
Also, if we really want to check this, shouldn't we look at all IDs in `mapped_rd_exact_set` instead of just one of them? Shouldn't we have the `break` at line 588?

Reply:
For a `mapped_rd` in `mapped_rd_exact_set`, we take its  `allConcretizedDomains` and store to `all_cids`.
The check is to check `all_cids`  are `fromSameExactGroup`.
The `break` at line 588 is to break from the loop over other `mapped_rd` in `mapped_rd_exact_set` as we already find a concretized domain, which can be used to represent the size of the broadcast root domain.

Reply:
>The break at line 588 is to break from the loop over other mapped_rd in mapped_rd_exact_set as we already find a concretized domain, which can be used to represent the size of the broadcast root domain.

That concretized domain is what `mapped_rd` is concretized into, but it isn't necessarily the same as what `rd` is concretized into. Note that exactly mapped broadcast IDs may be concretized into different IDs. 

Example:

```

TEST_F(NVFuserTest, TMP) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeSymbolicTensor(1);
  fusion.addInput(tv0);
  auto tv1 = makeSymbolicTensor(2);
  fusion.addInput(tv1);
  auto tv2 = makeSymbolicTensor(2);
  fusion.addInput(tv2);

  auto tv3 = broadcast(tv0, {false, true});

  auto tv4 = set(tv3);
  auto tv5 = add(tv4, tv1);

  auto tv6 = set(tv3);
  auto tv7 = add(tv6, tv2);

  fusion.addOutput(tv5);
  fusion.addOutput(tv7);

  fusion.printMath();

  ConcretizedBroadcastDomains concretized_bc(&fusion);

  std::cout << "tv3 concretization: "
            << toDelimitedString(
                   concretized_bc.allConcretizedDomains(tv3->axis(1)))
            << std::endl;
  std::cout << "tv4 concretization: "
            << toDelimitedString(
                   concretized_bc.allConcretizedDomains(tv4->axis(1)))
            << std::endl;
  std::cout << "tv6 concretization: "
            << toDelimitedString(
                   concretized_bc.allConcretizedDomains(tv6->axis(1)))
            << std::endl;

  ComputeAtMap ca_map(&fusion);
  std::cout << tv4->axis(1)->toString() << ", " << tv6->axis(1)->toString()
            << ": "
            << ca_map.areMapped(
                   tv4->axis(1), tv6->axis(1), IdMappingMode::EXACT)
            << std::endl;

  auto tv4_concretized_id =
      *(concretized_bc.allConcretizedDomains(tv4->axis(1)).begin());
  auto tv6_concretized_id =
      *(concretized_bc.allConcretizedDomains(tv6->axis(1)).begin());
  std::cout << tv4_concretized_id->toString() << ", "
            << tv6_concretized_id->toString()
            << ca_map.areMapped(
                   tv4_concretized_id, tv6_concretized_id, IdMappingMode::EXACT)
            << std::endl;
}
```
```
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from NVFuserTest
[ RUN      ] NVFuserTest.TMP
Inputs:
  T0_g[ iS0{i0} ], float
  T1_g[ iS1{i2}, iS2{i3} ], float
  T2_g[ iS3{i4}, iS4{i5} ], float
Outputs:
  T5_g[ iS9{i0}, iS10{i3} ], float
  T7_g[ iS13{i0}, iS14{i5} ], float

%kernel_math {
T3_l[ iS5{i0}, bS6{1} ]
   = broadcast( T0_g[ iS0{i0} ] )
T4_l[ iS7{i0}, bS8{1} ]
   = Set( T3_l[ iS5{i0}, bS6{1} ] )
T5_g[ iS9{i0}, iS10{i3} ]
   = T4_l[ iS7{i0}, bS8{1} ]
   + T1_g[ iS1{i2}, iS2{i3} ];
T6_l[ iS11{i0}, bS12{1} ]
   = Set( T3_l[ iS5{i0}, bS6{1} ] )
T7_g[ iS13{i0}, iS14{i5} ]
   = T6_l[ iS11{i0}, bS12{1} ]
   + T2_g[ iS3{i4}, iS4{i5} ];
}

tv3: iS14{i5}, iS10{i3}
tv4: iS10{i3}
tv6: iS14{i5}
bS8{1}, bS12{1}: 1
iS10{i3}, iS14{i5}0
```

The broadcast IDs of `tv4` and `tv6` are exactly mapped, but the IDs they are concretized into are not.

Reply:
If this happens, the fusion will be segmented by our heuristics, right? I added a test case `FusionAvoidRedundantWriteDifferentConcretizedDomains_CUDA` and tested pointwise and reduciton and the fusion is segmented in both cases. Just in case, I removed the `break` and added additional check:
```
            // make sure this broadcast root domain is concretized to the same domain
            // otherwise, treat as not concretized becase we don't know which one to use.
            auto iter = concretized_broadcast_root_domains_.find(rd);
            if(iter != concretized_broadcast_root_domains_.end() && iter->second != *all_cids.begin()) {
              concretized_broadcast_root_domains_.erase(iter);
              // break to move to the next broadcast root domain
              break;
            }else{
              concretized_broadcast_root_domains_[rd] = *all_cids.begin();
            }
```

Reply:
This is a part of fusion lowering, which is designed to be independent from how the schedulers and segmenter work, so we should not assume how they are segmented. In fact, we could just define a fusion like the example I posted above and pass it to `GpuLower`, and it should still work or at least error out if it cannot be handled without segmentation.

Reply:
Also, I still don't understand why concretization matters. As shown here, https://github.com/NVIDIA/Fuser/pull/11#issuecomment-1541141230, the issue happens even with non-concretized broadcast domains.

Reply:
> This is a part of fusion lowering, which is designed to be independent from how the schedulers and segmenter work, so we should not assume how they are segmented. In fact, we could just define a fusion like the example I posted above and pass it to `GpuLower`, and it should still work or at least error out if it cannot be handled without segmentation.

Added to test case `FusionAvoidRedundantWriteDifferentConcretizedDomains_CUDA`, where the non-segmented fusion is directly passed to `GpuLower`.

Reply:
> Also, I still don't understand why concretization matters. As shown here, [#11 (comment)](https://github.com/NVIDIA/Fuser/pull/11#issuecomment-1541141230), the issue happens even with non-concretized broadcast domains.

Marked as a case can't be handled by this PR.